### DISCUSSION TITLE
postgres: filter with elementwise array comparison

### DISCUSSION THREAD

**Discussion by fishfacegit** [sqlalchemy/sqlalchemy#11100] [Usage Questions] [answered]
### Describe the bug

filter with indexed elementwise array comparison does not work as expected.
It does not return the expected values.
`.filter(BlockMinHashLSH.b14r9 == lsh.b14r9)` returns 3 values, which is correct
` .filter(BlockMinHashLSH.b14r9[1] == lsh.b14r9[1])` returns nothing, should also return 3 values

### SQLAlchemy Version in Use

2.0.27

### DBAPI (i.e. the database driver)

psycopg[binary]

### Database Vendor and Major Version

PostgreSQL 16

### Python Version

3.12

### Operating system

Ubuntu / Linux

### To Reproduce

```python
q = (
            session.query(BASIC_BLOCK, BlockMinHashLSH)
            .distinct(BlockMinHashLSH.mnemonics_sha256_hash)
            .filter(
                BASIC_BLOCK.id != block.id,
                BlockMinHashLSH.b14r9.overlap(lsh.b14r9),
                BASIC_BLOCK.mnemonics_sha256_hash
                == BlockMinHashLSH.mnemonics_sha256_hash,
            )
            .filter(BlockMinHashLSH.b14r9 == lsh.b14r9)
            # .filter(BlockMinHashLSH.b14r9[1] == lsh.b14r9[1])
        )
candidates_lsh_list = q.all()
```

### Error

```
# Copy the complete stack trace and error message here, including SQL log output if applicable.
```

**Comment by CaselIT** [comment]
Hi,

There is no information on how to reproduce or what are the models mentioned.

Converting to discussion.

Can you share at least example data and example models/tables?

**Comment by zzzeek** [comment]
so SQLAlchemy doesnt search the database for rows, it renders SQL strings and sends them to your database.

therefore when reporting something like this, as the [bug template](https://github.com/sqlalchemy/sqlalchemy/blob/main/.github/ISSUE_TEMPLATE/bug_report.yaml) mentions, please include SQL output when reporting something unexpected.

in this case we want to look only at the SQL being rendered by the various statements you have.  is it what you expect?

**Comment by fishfacegit** [comment]
Sorry for the bad initial setup.
here i provide a working example for a postgres database. It is interesting that the last two testcases provide the same "query" but different outputs. It's probably related to postgres indexing starts at 1.


```py
import sys

import sqlalchemy as sa
from sqlalchemy import Index
from sqlalchemy.dialects.postgresql import ARRAY, BYTEA, INTEGER
from sqlalchemy.orm import Mapped, mapped_column

sys.path.append("..")

from database.tables.base import Base
from database.util import CONNECT_STRING


class TestTable(Base):
    __tablename__ = "test_table"

    sha256_hash: Mapped[bytes] = mapped_column(
        BYTEA(32),
        nullable=False,
        primary_key=True,
    )
    hashlist: Mapped[list[int]] = mapped_column(
        ARRAY(INTEGER()), nullable=True
    )

    __table_args__ = (
        Index(
            f"ix_{__tablename__}_hashlist",
            hashlist,
            postgresql_using="gin",
        ),
    )

engine = sa.create_engine(CONNECT_STRING)
Base.metadata.create_all(engine)

from sqlalchemy.orm import Session

data = [TestTable(sha256_hash=bytes([i] * 32), hashlist=list(range(9))) for i in range(3)]
print(f"comparison data: {data[0].hashlist[8]}")
with Session(engine) as session:
    session.add_all(data)
    # this works
    q = session.query(TestTable).filter(TestTable.hashlist == data[0].hashlist)
    print(q)
    print(q.all())
    print()
    # this works
    q = session.query(TestTable).filter(TestTable.hashlist[9] == data[0].hashlist[8])
    print(q)
    print(q.all())
    print()
    # this fails
    cond = sa.or_(*[x == y for x,y in zip(TestTable.hashlist, data[0].hashlist)])
    q = session.query(TestTable).filter(cond)
    print(q)
    print(q.all())
    print()
    # this fails
    cond = sa.or_(*[TestTable.hashlist[i] == data[0].hashlist[i] for i in range(9)])
    q = session.query(TestTable).filter(cond)
    print(q)
    print(q.all())
    print()
    # this works
    cond = sa.or_(*[TestTable.hashlist[i+1] == data[0].hashlist[i] for i in range(9)])
    q = session.query(TestTable).filter(cond)
    print(q)
    print(q.all())
```
```sql
comparison data: 8
SELECT test_table.sha256_hash AS test_table_sha256_hash, test_table.hashlist AS test_table_hashlist 
FROM test_table 
WHERE test_table.hashlist = %(hashlist_1)s::INTEGER[]
**[<__main__.TestTable object at 0x7ff27434d7f0>, <__main__.TestTable object at 0x7ff24fb4a3c0>, <__main__.TestTable object at 0x7ff264cc3ce0>]**

SELECT test_table.sha256_hash AS test_table_sha256_hash, test_table.hashlist AS test_table_hashlist 
FROM test_table 
WHERE test_table.hashlist[%(hashlist_1)s::INTEGER] = %(param_1)s::INTEGER
**[<__main__.TestTable object at 0x7ff27434d7f0>, <__main__.TestTable object at 0x7ff24fb4a3c0>, <__main__.TestTable object at 0x7ff264cc3ce0>]**

SELECT test_table.sha256_hash AS test_table_sha256_hash, test_table.hashlist AS test_table_hashlist 
FROM test_table 
WHERE test_table.hashlist[%(hashlist_1)s::INTEGER] = %(param_1)s::INTEGER OR test_table.hashlist[%(hashlist_2)s::INTEGER] = %(param_2)s::INTEGER OR test_table.hashlist[%(hashlist_3)s::INTEGER] = %(param_3)s::INTEGER OR test_table.hashlist[%(hashlist_4)s::INTEGER] = %(param_4)s::INTEGER OR test_table.hashlist[%(hashlist_5)s::INTEGER] = %(param_5)s::INTEGER OR test_table.hashlist[%(hashlist_6)s::INTEGER] = %(param_6)s::INTEGER OR test_table.hashlist[%(hashlist_7)s::INTEGER] = %(param_7)s::INTEGER OR test_table.hashlist[%(hashlist_8)s::INTEGER] = %(param_8)s::INTEGER OR test_table.hashlist[%(hashlist_9)s::INTEGER] = %(param_9)s::INTEGER
**[]**

SELECT test_table.sha256_hash AS test_table_sha256_hash, test_table.hashlist AS test_table_hashlist 
FROM test_table 
WHERE test_table.hashlist[%(hashlist_1)s::INTEGER] = %(param_1)s::INTEGER OR test_table.hashlist[%(hashlist_2)s::INTEGER] = %(param_2)s::INTEGER OR test_table.hashlist[%(hashlist_3)s::INTEGER] = %(param_3)s::INTEGER OR test_table.hashlist[%(hashlist_4)s::INTEGER] = %(param_4)s::INTEGER OR test_table.hashlist[%(hashlist_5)s::INTEGER] = %(param_5)s::INTEGER OR test_table.hashlist[%(hashlist_6)s::INTEGER] = %(param_6)s::INTEGER OR test_table.hashlist[%(hashlist_7)s::INTEGER] = %(param_7)s::INTEGER OR test_table.hashlist[%(hashlist_8)s::INTEGER] = %(param_8)s::INTEGER OR test_table.hashlist[%(hashlist_9)s::INTEGER] = %(param_9)s::INTEGER
**[]**

SELECT test_table.sha256_hash AS test_table_sha256_hash, test_table.hashlist AS test_table_hashlist 
FROM test_table 
WHERE test_table.hashlist[%(hashlist_1)s::INTEGER] = %(param_1)s::INTEGER OR test_table.hashlist[%(hashlist_2)s::INTEGER] = %(param_2)s::INTEGER OR test_table.hashlist[%(hashlist_3)s::INTEGER] = %(param_3)s::INTEGER OR test_table.hashlist[%(hashlist_4)s::INTEGER] = %(param_4)s::INTEGER OR test_table.hashlist[%(hashlist_5)s::INTEGER] = %(param_5)s::INTEGER OR test_table.hashlist[%(hashlist_6)s::INTEGER] = %(param_6)s::INTEGER OR test_table.hashlist[%(hashlist_7)s::INTEGER] = %(param_7)s::INTEGER OR test_table.hashlist[%(hashlist_8)s::INTEGER] = %(param_8)s::INTEGER OR test_table.hashlist[%(hashlist_9)s::INTEGER] = %(param_9)s::INTEGER
**[<__main__.TestTable object at 0x7ff27434d7f0>, <__main__.TestTable object at 0x7ff24fb4a3c0>, <__main__.TestTable object at 0x7ff264cc3ce0>]**
```

**Reply by CaselIT** [reply]
This all seems expected. I'm not sure what sqlalchemy should do differently here.

I honestly did not think this would work, but I guess it does? 
`sa.or_(*[x == y for x,y in zip(TestTable.hashlist, data[0].hashlist)])`
Seems that python just calls `__getitem__` to generate an iterator here.

**Reply by fishfacegit** [reply]
No, it does not work.
 `sa.or_(*[TestTable.hashlist[i] == data[0].hashlist[i] for i in range(9)])`
`cond = sa.or_(*[x == y for x,y in zip(TestTable.hashlist, data[0].hashlist)])`
sqlalchemy should build a query that returns the correct matching of items, or at least a warning.
If someone tries to create a dynamic query for multiple backends using sqlalchemy, the same code should result in the same outcome.
On the other hand this works as a workaround, but shouldn't because it is misleading:
`sa.or_(*[TestTable.hashlist[i+1] == data[0].hashlist[i] for i in range(9)])`
compares internally idx(1,1) ... idx(9,9) 
and not idx(1,0) ... (9,8)

**Reply by CaselIT** [reply]
I disagree. I'm not sure how are you expecting it to work?

How does sqlalchemy know that you are using a range vs a fixed value? Should `TestTable.hashlist[3]` actually mean 4?
That makes no sense and would be hugely surprising.

Overall this is issue is impossible to fix since postgresql decided to use 1 based indexing. Overall I think sqlalchemy is doing the least surprisingly thing here using the number you are giving it.

**Reply by fishfacegit** [reply]
`How does sqlalchemy know that you are using a range vs a fixed value?`
It does not matter if it's a range or a fixed value.
In python i would write TestTable.hashlist[3], and think that i am accessing the 4th element.
Since python is 0 indexed and postgres is 1 indexed, i would sqlalchemy expect to resolve this internally:
So that a condition that contains the first element ([0] in python) would match somehow the first element in postgres.
Right now it is shifted by 1 position, and might lead to unexpected results.

I think we disagree on this position:
```
Should TestTable.hashlist[3] actually mean 4?
That makes no sense and would be hugely surprising.
Overall I think sqlalchemy is doing the least surprisingly thing here using the number you are giving it.
```

**Reply by CaselIT** [reply]
@zzzeek do you have an opinion here? As stated I think not changing the values is the least surprising thing to do.

It seems that PG returns null for any index that does not exists, and does no validation (you could ask -100 too and get NULL).
Maybe, we could raise/warn while compiling if we get a value that's <= 0 since those have no meaning on pg?

(maybe just a warning since there may be pg derivate that use the python semantic wrt negative indexes)

**Comment by zzzeek** [comment, marked as answer]
hi -

there seems to be an issue in how Sphinx is building the documentation, that is shoving in a whole doc for the "Comparator" class above the init method, making the arguments for ARRAY difficult to find.

Because of that, you may have missed the arguments that are available.

in this case, set [zero_indexes](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#sqlalchemy.dialects.postgresql.ARRAY.params.zero_indexes) to True to get zero-indexed behavior

```
    hashlist: Mapped[list[int]] = mapped_column(
        ARRAY(INTEGER(), zero_indexes=True), nullable=True
    )

```

**Reply by zzzeek** [reply]
ive pushed some changes to the docs so `__init__` is more prominent and added extra notes referring to the zero_index parameter.

**Reply by CaselIT** [reply]
never noticed that parameter

**Reply by fishfacegit** [reply]
That's great, i didn't know that it was already implemented and could be triggered by the zero_indexes keyword.
